### ISSUE TITLE
MD046 - Triggers violation incorrectly

### ISSUE THREAD

**Issue by amimas** [DavidAnson/markdownlint#252]
I've been using [markdownlint-cli](https://github.com/igorshubovych/markdownlint-cli) in my docs repositories. I upgraded from version `0.15.0` to latest (`0.21.0`), which makes the rule *MD046* available.

This rule seems to be triggering violations incorrectly. Below is a snippet of content. Please ignore the `\` at the beginning of code block fences. I added those so that the entire snippet below gets rendered together on github.

```markdown

* Edit `path/to/file.xml` and add the following line beneath `<!-- Add new files`:

\```xml
<file>META-INF/filename.xml</file>
\```

* Create a new interface in `src/main/java` called `com.example.tc.TermsAndConditionService`.

* Add the following method definitions to the interface: _(**Hint:** don't forget to define each method with javadoc comments)_

\```java
a snippet of java code here
\```

```

MD046 rule is complaining about the following line:

```markdown
* Create a new interface in `src/main/java` called `com.example.tc.TermsAndConditionService`.
```

The error message is:

```sh
$ markdownlint ../docs/ ./versioned_docs/
../docs/path/to/doc.md:43 MD046/code-block-style Code block style [Expected: fenced; Actual: indented]
```

Notice that the line it is reporting problem does not have code blocks. It only has inline code wrapped with backticks. I'm getting similar errors being reported on a lot of other lines.

I've checked the existing issues related to MD046 but they didn't seem to mention the type of issues I've explained here. Hopefully this is not a duplicate issue.

**Comment by DavidAnson** [comment]
Is the problem still present if you remove the opening HTML comment (“<!--“) on the first line?

**Comment by amimas** [comment]
I think so because this is being reported on other files as well. I have since reverted back to the older version I was using. I'll have to try to reproduce the issue.

**Comment by DavidAnson** [comment]
If you can link to a file that shows this problem, I can narrow things down to identify the cause. Thanks!

**Comment by DavidAnson** [comment]
The example above does not seem to generate any errors in the current version of the library. Please provide an example that does so I can investigate further.

https://dlaa.me/markdownlint/#%25m%23%20Issue%20252%0A%0A*%20Edit%20%60path%2Fto%2Ffile.xml%60%20and%20add%20the%20following%20line%20beneath%20%60%3C!--%20Add%20new%20files%60%3A%0A%0A%60%60%60xml%0A%3Cfile%3EMETA-INF%2Ffilename.xml%3C%2Ffile%3E%0A%60%60%60%0A%0A*%20Create%20a%20new%20interface%20in%20%60src%2Fmain%2Fjava%60%20called%20%60com.example.tc.TermsAndConditionService%60.%0A%0A*%20Add%20the%20following%20method%20definitions%20to%20the%20interface%3A%20_(**Hint%3A**%20don't%20forget%20to%20define%20each%20method%20with%20javadoc%20comments)_%0A%0A%60%60%60java%0Aa%20snippet%20of%20java%20code%20here%0A%60%60%60

**Comment by amimas** [comment]
Okay. I'll try to get a working sample. Thanks for looking into this.

**Comment by DavidAnson** [comment]
Thanks! FYI, the latest version of the CLI is now 0.22.0.

**Comment by amimas** [comment]
I just had a chance to debug this a little bit more and I realized that the error actually comes from few lines of texts above what I had originally posted. I also tried with the latest version and seem to be having the same issue.

I used the online tool you posted above, with an extended snippet of my content and was able to reproduce it there; [reproducible issue on markdownlint demo](https://dlaa.me/markdownlint/#%25m%23%20Issue%20252%0A%0A*%20Add%20the%20following%20named%20query%20elements%20within%20the%20%60%3Centity-mappings%3E%60%20element%20of%20the%20new%20%60tc.xml%60%3A%0A%0A%60%60%60xml%0A%3Cnamed-query%20name%3D%22FIND_TNC_BY_CODE%22%3E%0A%20%20%20%20%3Cquery%3E%0A%20%20%20%20%20%20%20%20SELECT%20termsAndCondition%0A%20%20%20%20%20%20%20%20FROM%20TermsAndConditionsFlagImpl%20AS%20termsAndCondition%0A%20%20%20%20%20%20%20%20WHERE%20termsAndCondition.code%20%3D%0A%20%20%20%20%3C%2Fquery%3E%0A%3C%2Fnamed-query%3E%0A%0A%3Cnamed-query%20name%3D%22FIND_ALL_TNCS%22%3E%0A%20%20%20%20%3Cquery%3E%0A%20%20%20%20%20%20%20%20SELECT%20termsAndCondition%0A%20%20%20%20%20%20%20%20FROM%20TermsAndConditionsFlagImpl%20AS%20termsAndCondition%0A%20%20%20%20%3C%2Fquery%3E%0A%3C%2Fnamed-query%3E%0A%60%60%60%0A%0A*%20Edit%20%60path%2Fto%2Ffile.xml%60%20and%20add%20the%20following%20line%20beneath%20%60%3C!--%20Add%20new%20mapping%20files%60%3A%0A%0A%60%60%60xml%0A%3Cfile%3EMETA-INF%2Ffilename.xml%3C%2Ffile%3E%0A%60%60%60%0A%0A*%20Create%20a%20new%20interface%20in%20%60src%2Fmain%2Fjava%60%20called%20%60com.example.tc.TermsAndConditionsFlagService%60.%0A)

Also, while preparing this snippet, I noticed markdownlint is reporting the following error, which also seems to be false positive.

```log
28 - MD032 Lists should be surrounded by blank lines [Context: "\"]
```

The above error is complaining about not having blank lines after a list item at the end of the contents. But clearly there is an empty line. If I add an extra empty line (2 blank lines at the end), then the above error is not reported anymore.

**Comment by alexfornuto** [comment]
Thanks for the great plugin! I'm also getting a conflict with MD046 and MD007 in VSCode. A bulleted list under a numbered step needs a 4-space indent, which MD046 is seeing as an indented code block. It's the inverse of @amimas's issue, since in my example the bullets come before the code fences.

I reckon the simple fix is to have MD046 ignore tabbed lines that start with a single bullet character followed by a space.

**Comment by DavidAnson** [comment]
Could you please provide an example of your scenario? I’ll make sure both scenarios are covered by the fix.

**Comment by alexfornuto** [comment]
Sure! As above, I'm adding `\` to my code fences for GitHub comment formatting:

```
1. The **Google IdP Information** modal provides values and a certificate to pass back to the WP SAML Auth plugin. The key details are in the Option 1 section:

  ![Google IdP Information](../../../images/guides/wordpress-google-sso/google-idp-information.png)

    - **SSO URL**: URL for WordPress to redirect to when initiating the SSO process.

    - **Entity ID**: How WordPress knows to recognize Google Apps.

    - **Certificate** - How Google Apps knows to trust a request coming from WordPress.

1. Copy the **SSO URL** and **Entity ID** values and paste them into the corresponding fields under **Identity Provider Settings** on the plugin settings page:

  ![Values from the Google SAML App copied to the WP SAML Auth settings](../../../images/guides/wordpress-google-sso/plugin-idp-values.png)

1. Download and commit the certificate file to your [`private` directory](/private-paths#private-path-for-code) so it's inaccessible from the web. You can add upload this file over [sftp](/sftp), or commit it to a local clone of the git repository, then commit and push. If the latter, you will need to first switch the connection mode to git:


  \```bash{outputLines: 2-6, 8}
  sftp -o Port=2222 google-saml.1f2a347b....0@appserver.google-saml.1f2a347b....drush.in
  sftp> cd code/private/
  sftp> put GoogleIDPCertificate-mydomain.com.pem
  Uploading GoogleIDPCertificate-mydomain.com.pem to /srv/bindings/972dd.../code/private/GoogleIDPCertificate-mydomain.com.pem
  GoogleIDPCertificate-alexfornuto.com.pem                                                                            100% 1253    27.6KB/s   00:00
  sftp> quit
  terminus env:commit --message "Added the Google IDP Certificate to a protected path" $site.$env
   [notice] Your code was committed.
  \```
```

**Comment by DavidAnson** [comment]
@alexfornuto, there only need to be three spaces for the sub-list. After making that change and tweaking some of the other spacing, your example does not report any errors with the currently-published `markdownlint` implementation:

https://dlaa.me/markdownlint/#%25m%23%20Issue%20252%0A%0A1.%20The%20**Google%20IdP%20Information**%20modal%20provides%20values%20and%20a%20certificate%20to%20pass%20back%20to%20the%20WP%20SAML%20Auth%20plugin.%20The%20key%20details%20are%20in%20the%20Option%201%20section%3A%0A%0A%20%20%20!%5BGoogle%20IdP%20Information%5D(..%2F..%2F..%2Fimages%2Fguides%2Fwordpress-google-sso%2Fgoogle-idp-information.png)%0A%0A%20%20%20-%20**SSO%20URL**%3A%20URL%20for%20WordPress%20to%20redirect%20to%20when%20initiating%20the%20SSO%20process.%0A%0A%20%20%20-%20**Entity%20ID**%3A%20How%20WordPress%20knows%20to%20recognize%20Google%20Apps.%0A%0A%20%20%20-%20**Certificate**%20-%20How%20Google%20Apps%20knows%20to%20trust%20a%20request%20coming%20from%20WordPress.%0A%0A1.%20Copy%20the%20**SSO%20URL**%20and%20**Entity%20ID**%20values%20and%20paste%20them%20into%20the%20corresponding%20fields%20under%20**Identity%20Provider%20Settings**%20on%20the%20plugin%20settings%20page%3A%0A%0A%20%20%20!%5BValues%20from%20the%20Google%20SAML%20App%20copied%20to%20the%20WP%20SAML%20Auth%20settings%5D(..%2F..%2F..%2Fimages%2Fguides%2Fwordpress-google-sso%2Fplugin-idp-values.png)%0A%0A1.%20Download%20and%20commit%20the%20certificate%20file%20to%20your%20%5B%60private%60%20directory%5D(%2Fprivate-paths%23private-path-for-code)%20so%20it's%20inaccessible%20from%20the%20web.%20You%20can%20add%20upload%20this%20file%20over%20%5Bsftp%5D(%2Fsftp)%2C%20or%20commit%20it%20to%20a%20local%20clone%20of%20the%20git%20repository%2C%20then%20commit%20and%20push.%20If%20the%20latter%2C%20you%20will%20need%20to%20first%20switch%20the%20connection%20mode%20to%20git%3A%0A%0A%20%20%20%60%60%60bash%7BoutputLines%3A%202-6%2C%208%7D%0A%20%20%20sftp%20-o%20Port%3D2222%20google-saml.1f2a347b....0%40appserver.google-saml.1f2a347b....drush.in%0A%20%20%20sftp%3E%20cd%20code%2Fprivate%2F%0A%20%20%20sftp%3E%20put%20GoogleIDPCertificate-mydomain.com.pem%0A%20%20%20Uploading%20GoogleIDPCertificate-mydomain.com.pem%20to%20%2Fsrv%2Fbindings%2F972dd...%2Fcode%2Fprivate%2F%20GoogleIDPCertificate-mydomain.com.pem%0A%20%20%20GoogleIDPCertificate-alexfornuto.com.%20pem%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20100%25%201253%20%20%20%2027.6KB%2Fs%20%20%2000%3A00%0A%20%20%20sftp%3E%20quit%0A%20%20%20terminus%20env%3Acommit%20--message%20%22Added%20the%20Google%20IDP%20Certificate%20to%20a%20protected%20path%22%20%24site.%24env%0A%20%20%20%20%5Bnotice%5D%20Your%20code%20was%20committed.%0A%20%20%20%60%60%60%0A

**Comment by amimas** [comment]
> there only need to be three spaces for the sub-list. 

@DavidAnson - Thanks for looking into this.  Noticed you added the label `fixed in next`. I think I have a lot of contents that has sub-list items and they are indented with 4 spaces. Is that also going to work in that next version?

**Comment by DavidAnson** [comment]
Yes, the fix was to ignore un-closed HTML comments. All formatting will be handled the same.

**Comment by alexfornuto** [comment]
@DavidAnson The problem with using 3 spaces to indent the sub-list is that it triggers MD007:
![image](https://user-images.githubusercontent.com/2349184/76359134-a47fe000-62e8-11ea-829d-be6f3cadc606.png)

**Comment by DavidAnson** [comment]
Please open my example link above. That is your scenario. I used 3 spaces for that list without warnings. I think I also had to invent the image above it by 3 spaces.

**Comment by alexfornuto** [comment]
Sorry, tried to update my comment before you saw it.

**Comment by DavidAnson** [comment]
Cool. The general guideline I use is that content of a list item should be indented to match the list marker. That is usually 2 spaces for unordered lists and 3 spaces for ordered lists.

The spec has a lot more detail: https://spec.commonmark.org/0.29/#list-items

**Comment by amimas** [comment]
@DavidAnson - Thanks again for looking into this quickly. Any idea when this will be released? In my projects, I don't really consume this directly. It's a transitive dependency through markdownlint-cli.

**Comment by DavidAnson** [comment]
Maybe 1 week for the library and 2 more for the CLI?
